### PR TITLE
Parameterize result-format for R array UDFs

### PR DESCRIPTION
**Pull request by johnkerl**
# Context

As part of ongoing work to bring R array UDFs up to feature-parity with Python array UDFs, here we honor the `result_format` parameter.

# Validation

## Input

```
#!/usr/bin/env Rscript

library(tiledbcloud)
library(tiledb)

# ----------------------------------------------------------------
username <- Sys.getenv("TCR_USERNAME", unset="demo")
password <- Sys.getenv("TCR_PASSWORD", unset="demodemo")
host     <- Sys.getenv("TCR_HOST",     unset="http://localhost:8181")
tiledbcloud::login(username=username, password=password, host=host)
# ----------------------------------------------------------------

namespace <- "demo"
uri       <- "quickstart_dense"

myfunc <- function(df) {
  vec <- as.vector(df[["a"]])
  list(min=min(vec), med=median(vec), max=max(vec))
}

selectedRanges <- list(cbind(1,2), cbind(1,2))
#selectedRanges <- list(rows=cbind(1,2), cols=cbind(3,4))

cat("DEFAULT RESULT_FORMAT\n")
result <- tiledbcloud::execute_array_udf(
  namespace=namespace,
  array=uri,
  udf=myfunc,
  selectedRanges=selectedRanges
)
print(result)

cat("\n")
cat("NATIVE RESULT_FORMAT\n")
result <- tiledbcloud::execute_array_udf(
  namespace=namespace,
  array=uri,
  udf=myfunc,
  result_format='native',
  selectedRanges=selectedRanges
)
print(result)

cat("\n")
cat("JSON RESULT_FORMAT\n")
result <- tiledbcloud::execute_array_udf(
  namespace=namespace,
  array=uri,
  udf=myfunc,
  result_format='json',
  selectedRanges=selectedRanges
)
print(result)

cat("\n")
cat("ARROW RESULT_FORMAT\n")
result <- tiledbcloud::execute_array_udf(
  namespace=namespace,
  array=uri,
  udf=myfunc,
  result_format='arrow',
  selectedRanges=selectedRanges
)
print(result)
```

## Output

```
DEFAULT RESULT_FORMAT
$min
[1] 1

$med
[1] 3.5

$max
[1] 6

NATIVE RESULT_FORMAT
$min
[1] 1

$med
[1] 3.5

$max
[1] 6

JSON RESULT_FORMAT
     type value
1 integer     1
2  double   3.5
3 integer     6

ARROW RESULT_FORMAT
  min med max
1   1 3.5   6
```

### CODE REVIEW

File: R/execute.R [hunk 1]
```diff
@@ -95,29 +98,47 @@ execute_array_udf <- function(namespace, array, udf, selectedRanges, layout=NULL
...
+    native={
+      formattedResult <- unserialize(result, NULL)
+    },
+    json={
+      resultString <- rawToChar(result)
+      resultJSON <- jsonlite::fromJSON(resultString)
+      formattedResult <- resultJSON$value
+    },
+    arrow={
+      formattedResult <- arrow::read_ipc_stream(result)
```
Comment (johnkerl) [inline R/execute.R:138]: Decision point: @aaronwolen @eddelbuettel do we want to decode here, or return the result as-is to the calling script so they can `arrow::read_ipc_stream` as they wish?

Comment (eddelbuettel) [reply R/execute.R:138]: Who is "they" here?

**Review by eddelbuettel — APPROVED** [review]

File: R/execute.R (hunk 1 above)
Comment (johnkerl) [reply R/execute.R:138]: Sorry -- people using `TileDB-Cloud-R`

**Review by aaronwolen — APPROVED** [review]

**Comment by eddelbuettel** [comment]
These two make no sense to me:

```
DEFAULT RESULT_FORMAT
     type value
1 integer     1
2  double   3.5
3 integer     6

JSON RESULT_FORMAT
     type value
1 integer     1
2  double   3.5
3 integer     6
```

By the time they are assigned the type is redundant.  Yet the column name and 'what called them' is missing.  Can that be added or is it lost in translation?

File: R/execute.R [hunk 2]
```diff
@@ -74,7 +74,7 @@ execute_generic_udf <- function(namespace, udf, args=NULL) {
 ##' @importFrom arrow read_ipc_stream
 ##'
 ##' @export
-execute_array_udf <- function(namespace, array, udf, selectedRanges, layout=NULL, result_format='json', attrs=NULL) {
+execute_array_udf <- function(namespace, array, udf, selectedRanges, layout=NULL, result_format='native', attrs=NULL) {
```
Comment (eddelbuettel) [inline R/execute.R:77]: :100:

File: R/execute.R (hunk 1 above)
Comment (johnkerl) [reply R/execute.R:152]: For the record: off-line conversation with @Shelnutt2 : this parameter indeed controls the wire-protocol used between the REST server and this library; in all cases the return value back to the user should be deserialized -- they do not receive JSON or Arrow. This is intentional and consistent with the `TileDB-Cloud-Py` library. If we decide to later add a flag allowing return-to-caller of JSON or Arrow, that will be a separate feature, implemented for Python as well as R.

**Comment by johnkerl** [comment]
Note: fix for (unrelated) `openssl` CI error coming soon

File: DESCRIPTION [hunk 3]
```diff
@@ -14,6 +14,6 @@ Depends: R (>= 3.3)
 License: MIT + file LICENSE
 LazyData: true
 Suggests: testthat, tinytest
-Imports: jsonlite, httr, R6, base64enc
+Imports: jsonlite, arrow, httr, R6, base64enc
```
Comment (eddelbuettel) [inline DESCRIPTION:17]: Just FYI for the CI depends:  arrow is both somewhat expensive to build, and somewhat fragile.  Maybe keep an eye on this one.

Comment (johnkerl) [reply DESCRIPTION:17]: Awesome context -- thanks @eddelbuettel !! :D

**Pull request merged**